---
title: "Extended Split-ORFs analysis report"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
params:
    args: myarg
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```


```{r imports, include = FALSE}
library(seqinr)
library(ggplot2)
library(geomtextpath)
library(tidyr)
library(stringr)
library(dplyr)
library(VennDiagram)
library(RColorBrewer)
library(readr)
library(grid)
```


This report shows a number of summary statistics of the a result files obtained from the Extended Split-ORFs Pipeline.

## DNA uniqueness

### Overview of the found unique DNA regions:

```{r,echo=FALSE,message=FALSE,results='asis'}
datapath=getwd()
file1=paste0(datapath,params[[1]][[1]])
file2=paste0(datapath,params[[1]][[2]])
file3=paste0(datapath,params[[1]][[3]])
file4=paste0(datapath,params[[1]][[4]])
file5=paste0(datapath,params[[1]][[5]])
file6 = paste0(datapath,params[[1]][[6]])
file7 = paste0(datapath,params[[1]][[7]])
file8 = paste0(datapath,params[[1]][[8]])
file9 = paste0(datapath,params[[1]][[9]])
file10 = paste0(datapath,params[[1]][[10]])
file11 = paste0(datapath,params[[1]][[11]])

# file1 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_DNA_regions.fa'
# file2 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_Protein_regions.fa'
# file3 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_DNA_Regions_gt20_filtered.bed'
# file4 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_Protein_Regions_gt8_valid_filtered.bed'
# file5 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/UniqueProteinORFPairs.txt'
# file6 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_DNA_Regions_for_riboseq.bed'
# file7 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_Protein_Regions_transcript_coords.bed'
# file8 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_DNA_regions_genomic.bed'
# file9 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_Protein_regions_genomic.bed'
# file10 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_16.12.2024-10.20.59_NMD_genomic_correction/Unique_Regions_Overlap_transcriptomic.bed'
# file11 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_28.11.2024-10.28.44_NMD_genomic_regions_with_unique_trans_coords/Unique_Regions_Overlap.bed'

# file1 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_11.09.2024-11.12.39_blast_all_NMD_ref_filtered_non_overlapping/Unique_DNA_regions.fa'
# file2 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_11.09.2024-11.12.39_blast_all_NMD_ref_filtered_non_overlapping/Unique_Protein_regions.fa'
# file3 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_11.09.2024-11.12.39_blast_all_NMD_ref_filtered_non_overlapping/Unique_DNA_Regions_gt20_filtered.bed'
# file4 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_11.09.2024-11.12.39_blast_all_NMD_ref_filtered_non_overlapping/Unique_Protein_Regions_gt8_valid_filtered.bed'
# file5 = '/Users/christina/Documents/ExtendedSplitORFPipeline-master_13_11_23/Output/run_11.09.2024-11.12.39_blast_all_NMD_ref_filtered_non_overlapping/UniqueProteinORFPairs.txt'



dnaresult<-read.fasta(file1)
cat(paste0("Number of Unique Regions:\t",length(dnaresult),"  \n"))
list=getSequence(dnaresult,as.string = TRUE)
lengthlist<- rep(NA, length(list))
for(i in 1:length(list)){
  lengthlist[i]=nchar(list[[i]][[1]])
}
cat(paste0("Average length of unique region:\t",round(mean(lengthlist)),"  \n"))
cat(paste0("Median length of unique region:\t",median(lengthlist),"  \n"))
cat(paste0("Maximal length of unique region:\t",max(lengthlist)," (for region ", substr(getName(dnaresult[match(max(lengthlist),lengthlist)]),start = 1,stop = 31), ")  \n"))
DNA_ids=substr(getName(dnaresult),start = 1,stop = 31)
DNA_Unique_per_transcript=as.vector(table(DNA_ids))


DNA_bed <- as.data.frame(read.table(file3,header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))
protein_bed <- as.data.frame(read.table(file4,header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))


unique_split_orfs = read.table(file5,header=T)
```


### This is the length distribution of the unique DNA regions

```{r, echo=FALSE}
dna_lengths <- data.frame(dna_ur_len = unlist(lengthlist))
#hist(lengthlist,breaks=100,col="blue",main="Length distribution of unique regions",xlab="Length of unique region")
ggplot(dna_lengths, aes(x=dna_ur_len)) + geom_histogram(bins=100) +
  labs(title="Length distribution of DNA unique regions", x="Length of DNA unique region", y="Frequency")+
  theme_minimal()
```
```{r, echo=FALSE}
ggplot(dna_lengths, aes(x=dna_ur_len)) + geom_histogram(bins = 100) +
  labs(title="Length distribution of DNA unique regions - zoom", x="Length of DNA unique region", y="Frequency")+
  theme_minimal()+ xlim(0, 500)
```




### This is the distribution of the unique DNA regions per transcript

```{r, echo=FALSE}
DNA_un_trans <- data.frame(dna_per_trans = DNA_Unique_per_transcript)
ggplot(DNA_un_trans, aes(x=dna_per_trans)) + geom_histogram(bins = 100) +
  labs(title="Number of unique DNA regions per transcript", x="Number of unique DNA regions", y="Frequency")+
  theme_minimal()
#hist(DNA_Unique_per_transcript,breaks=100,col="blue",main="Unique DNA regions per transcript",xlab="number of unique regions")
```
### Position of unique DNA regions within transcripts
```{r, echo = FALSE}
#position is still with respect to the ORF
DNA_bed$OrfTransID <- str_split_fixed(str_split_fixed(DNA_bed$V1, '\\|', 2)[,2], ':', 2)[,1]
DNA_bed$ORF_nr <- str_split_fixed(str_split_fixed(DNA_bed$V1, '\\|', 2)[,2], ':', 3)[,2]
DNA_bed$ORF_start <- as.numeric(str_split_fixed(DNA_bed$V1, ':', 4)[,3])
DNA_bed <- DNA_bed %>%
  rename(unique_start = V2)
DNA_bed <- DNA_bed %>%
  rename(unique_end = V3)
#DNA_bed
```

```{r, echo = FALSE}
#get ORF start and end positions
df_expanded <- separate_rows(unique_split_orfs, OrfPos, sep = ",")
ORF_start_end <- str_split_fixed(df_expanded$OrfPos, '-', 2)
df_expanded$ORF_start <- as.numeric(ORF_start_end[,1])
df_expanded$ORF_end <- as.numeric(ORF_start_end[,2])
#and ORf number inside of the transcript (according to start position)
df_expanded <- df_expanded %>%group_by(OrfTransID) %>%
  arrange(ORF_start) %>%
  mutate(Cumulative_ORF_Count = row_number())%>% 
  arrange(OrfTransID)

df_expanded$orf_number = "middle"
df_expanded[df_expanded$Cumulative_ORF_Count == 1,]$orf_number = "first"
max_ORF_rows <- df_expanded %>% group_by(OrfTransID) %>% slice(which.max(Cumulative_ORF_Count))
df_expanded[df_expanded$OrfPos %in% max_ORF_rows$OrfPos, ]$orf_number = "last"
```

```{r, echo = FALSE}
df_ORF_info_and_unique <- DNA_bed %>%
  left_join(df_expanded, by = c("OrfTransID", "ORF_start"))



df_ORF_info_and_unique$unique_start_percent <- df_ORF_info_and_unique$unique_start/(df_ORF_info_and_unique$ORF_end - df_ORF_info_and_unique$ORF_start)

df_ORF_info_and_unique$unique_end_percent <- df_ORF_info_and_unique$unique_end/(df_ORF_info_and_unique$ORF_end - df_ORF_info_and_unique$ORF_start)

```



```{r, echo=FALSE}
ggplot(df_ORF_info_and_unique, aes(x = unique_start_percent, fill = factor(orf_number))) +
  geom_histogram() +
    scale_fill_hue(l=45)+
  labs(title= "Start positions of unique regions inside of ORFs in %, by ORF position", x = "Start positions of unique regions inside of ORFs in %", y = "Frequency", fill = "unique region position in ORF") +
  theme_minimal()
```

```{r, echo=FALSE}
ggplot(df_ORF_info_and_unique, aes(x = unique_end_percent, fill = factor(orf_number))) +
  geom_histogram() +
    scale_fill_hue(l=45)+
  labs(title= "End positions of unique regions inside of ORFs in %, by ORF position", x = "End positions of unique regions inside of ORFs in %", y = "Frequency", fill = "unique region position in ORF") +
  theme_minimal()
```




```{r, echo=FALSE}
perfect_uniques_frame <- df_ORF_info_and_unique[(df_ORF_info_and_unique$unique_start_percent < 0.01 & df_ORF_info_and_unique$orf_number == 'last') | (df_ORF_info_and_unique$unique_end_percent > 0.99 & df_ORF_info_and_unique$orf_number == 'first'), ]
perfect_uniques_frame <- perfect_uniques_frame %>% group_by(OrfTransID) %>% summarize(count = n_distinct(orf_number))
perfect_uniques_frame <- perfect_uniques_frame[perfect_uniques_frame$count > 1,]
unique_regions_per_trans <- df_ORF_info_and_unique %>% group_by(OrfTransID) %>% summarize(count = n_distinct(ORF_nr), NumOrfs=first(NumOrfs))
unique_regions_at_least_2 <- unique_regions_per_trans[unique_regions_per_trans$count > 1, ]
unique_regions_for_each_ORF <- unique_regions_per_trans[unique_regions_per_trans$NumOrfs == unique_regions_per_trans$count, ]
cat(paste0("Number of transcripts with at least 2 unique DNA regions in 2 different ORFs:\t",length(unique_regions_at_least_2$count),"  \n"))
cat(paste0("Number of transcripts with 2 unique DNA regions at the end of first ORF and beginning last:\t",length(perfect_uniques_frame$OrfTransID),"  \n"))
cat(paste0("Number oftranscripts where all ORFs have unique regions:\t",length(unique_regions_for_each_ORF$NumOrfs),"  \n"))

```



## Protein uniqueness

### Overview of the found unique Protein regions:

```{r Protein stats,echo=FALSE,message=FALSE,results='asis'}
proteinresult<-read.fasta(file2)#"E:/Justin_Backup_29.04.2020/Justin/Uni-Frankfurt/FP+Masterarbeit/PipeTest/Pipeline/Output/clean_run/Unique_Protein_Regions.fa"
cat(paste0("Number of Unique Regions:\t",length(proteinresult),"  \n"))
protlist=getSequence(proteinresult,as.string = TRUE)
protlengthlist<- rep(NA, length(protlist))
for(i in 1:length(protlist)){
  protlengthlist[i]=nchar(protlist[[i]][[1]])
}
cat(paste0("Average length of unique region:\t",round(mean(protlengthlist)),"  \n"))
cat(paste0("Median length of unique region:\t",median(protlengthlist),"  \n"))
cat(paste0("Maximal length of unique region:\t",max(protlengthlist)," (for region ", getName(proteinresult[match(max(protlengthlist),protlengthlist)]), ")  \n"))
Protein_ids=substr(getName(proteinresult),start = 1,stop = 31)
Protein_Unique_per_transcript=as.vector(table(Protein_ids))
```

### This is the length distribution of the unique protein regions

```{r, echo=FALSE}
prot_lengths <- data.frame(prot_ur_len = unlist(protlengthlist))
ggplot(prot_lengths, aes(x=prot_ur_len)) + geom_histogram(bins=100) +
  labs(title="Length distribution of protein unique regions", x="Length of protein unique region", y="Frequency")+
  theme_minimal()
#hist(protlengthlist,breaks=100,col="blue",main="Length distribution of unique regions",xlab="Length of unique region")
```

```{r, echo=FALSE}
ggplot(prot_lengths, aes(x=prot_ur_len)) + geom_histogram(bins = 100) +
  labs(title="Length distribution of potein unique regions - zoom", x="Length of protein unique region", y="Frequency")+
  theme_minimal()+ xlim(0, 175)
```


### This is the distribution of the unique Protein regions per transcript

```{r, echo=FALSE}
prot_un_trans <- data.frame(prot_per_trans = Protein_Unique_per_transcript)
ggplot(prot_un_trans, aes(x=prot_per_trans)) + geom_histogram(bins = 100) +
  labs(title="Number of unique protein regions per transcript", x="Number of unique protein regions", y="Frequency")+
  theme_minimal()

#hist(Protein_Unique_per_transcript,breaks=100,col="blue",main="Unique Protein regions per transcript",xlab="number of unique regions")
```


### Position of unique protein regions within transcripts
```{r, echo = FALSE}
#position is still with respect to the ORF
protein_bed$OrfTransID <- str_split_fixed(str_split_fixed(protein_bed$V1, '\\|', 2)[,2], ':', 2)[,1]
protein_bed$ORF_nr <- str_split_fixed(str_split_fixed(protein_bed$V1, '\\|', 2)[,2], ':', 3)[,2]
protein_bed$ORF_start <- as.numeric(str_split_fixed(protein_bed$V1, ':', 4)[,3])
protein_bed <- protein_bed %>%
  rename(unique_start = V2)
protein_bed <- protein_bed %>%
  rename(unique_end = V3)
#DNA_bed
```


```{r, echo = FALSE}
df_ORF_info_and_unique_protein <- protein_bed %>%
  left_join(df_expanded, by = c("OrfTransID", "ORF_start"))



df_ORF_info_and_unique_protein$unique_start_percent <- df_ORF_info_and_unique_protein$unique_start/((df_ORF_info_and_unique_protein$ORF_end - df_ORF_info_and_unique_protein$ORF_start)/3)

df_ORF_info_and_unique_protein$unique_end_percent <- df_ORF_info_and_unique_protein$unique_end/((df_ORF_info_and_unique_protein$ORF_end - df_ORF_info_and_unique_protein$ORF_start)/3)

#df_ORF_info_and_unique

```


```{r, echo=FALSE}
ggplot(df_ORF_info_and_unique_protein, aes(x = unique_start_percent, fill = factor(orf_number))) +
  geom_histogram() +
    scale_fill_hue(l=45)+
  labs(title= "Start positions of unique protein regions inside of ORFs in %, by ORF position", x = "Start positions of unique protein regions inside of ORFs in %", y = "Frequency", fill = "unique protein region position in ORF") +
  theme_minimal()
```

```{r, echo=FALSE}
ggplot(df_ORF_info_and_unique_protein, aes(x = unique_end_percent, fill = factor(orf_number))) +
  geom_histogram() +
    scale_fill_hue(l=45)+
  labs(title= "End positions of unique protein regions inside of ORFs in %, by ORF position", x = "End positions of unique protein regions inside of ORFs in %", y = "Frequency", fill = "unique protein region position in ORF") +
  theme_minimal()
```



```{r, echo=FALSE}
perfect_uniques_frame <- df_ORF_info_and_unique_protein[(df_ORF_info_and_unique_protein$unique_start_percent < 0.01 & df_ORF_info_and_unique_protein$orf_number == 'last') | (df_ORF_info_and_unique_protein$unique_end_percent > 0.99 & df_ORF_info_and_unique_protein$orf_number == 'first'), ]
perfect_uniques_frame <- perfect_uniques_frame %>% group_by(OrfTransID) %>% summarize(count = n_distinct(orf_number))
perfect_uniques_frame <- perfect_uniques_frame[perfect_uniques_frame$count > 1,]
unique_regions_per_trans <- df_ORF_info_and_unique_protein %>% group_by(OrfTransID) %>% summarize(count = n_distinct(ORF_nr), NumOrfs=first(NumOrfs))
unique_regions_at_least_2 <- unique_regions_per_trans[unique_regions_per_trans$count > 1, ]
unique_regions_for_each_ORF <- unique_regions_per_trans[unique_regions_per_trans$NumOrfs == unique_regions_per_trans$count, ]
cat(paste0("Number of transcripts with at least 2 unique Protein regions in 2 different ORFs:\t",length(unique_regions_at_least_2$count),"  \n"))
cat(paste0("Number of transcripts with 2 unique Protein regions at the end of first ORF and beginning last:\t",length(perfect_uniques_frame$count),"  \n"))
cat(paste0("Number oftranscripts where all ORFs have unique regions:\t",length(unique_regions_for_each_ORF$NumOrfs),"  \n"))

```
## Overall statistics

```{r overall, echo=FALSE,results='asis'}
cat(paste0("Number of transcripts with unique DNA region:\t",length(unique(DNA_ids)),"  \n"))
cat(paste0("Number of transcripts with unique Protein region:\t",length(unique(Protein_ids)),"  \n"))
cat(paste0("Number of transcripts with unique DNA and Protein region:\t",length(intersect(DNA_ids,Protein_ids)),"  \n"))
```

# Overlap of unique protein and DNA regions
## transcriptomic
```{r}

DNA_regions_transcriptomic <- readr::read_tsv(file6, col_names = FALSE)
protein_regions_transcriptomic <- readr::read_tsv(file7, col_names = FALSE)


# Chart
venn.plot <- venn.diagram(
  x = list(unique(DNA_regions_transcriptomic$X4), unique(protein_regions_transcriptomic$X4)),
  category.names = c("ORFs with unique DNA regions", "ORFs with unique protein regions"),
  fill = c("skyblue", "lightpink"),
  lty = "blank",
  alpha = 0.5, 
  filename = NULL,
  # output=TRUE
  cat.pos = c(-27, 27)
)
grid.newpage()
grid::grid.draw(venn.plot)

grid.text("Venn Diagram of DNA and Protein Transcriptomic Unique Regions", 
          x = 0.5,            # Horizontal position (0 = left, 1 = right)
          y = 0.98,           # Vertical position (0 = bottom, 1 = top)
          gp = gpar(fontsize = 12, fontface = "bold"))  # Font settings
```
```{r}

DNA_regions_genomic <- readr::read_tsv(file8, col_names = FALSE)
protein_regions_genomic <- readr::read_tsv(file9, col_names = FALSE)

protein_region_ORFs = str_split_fixed(protein_regions_genomic$X4, ':', 3)[,2]
DNA_region_ORFs = str_split_fixed(DNA_regions_genomic$X4, ':', 3)[,2]

# Chart
venn.plot <- venn.diagram(
  x = list(unique(DNA_region_ORFs), unique(protein_region_ORFs)),
  category.names = c("ORFs with unique DNA regions", "ORFs with unique protein regions"),
  fill = c("skyblue", "lightpink"),
  lty = "blank",
  alpha = 0.5, 
  filename = NULL,
  # output=TRUE
  cat.pos = c(-27, 27)
)
grid.newpage()
grid::grid.draw(venn.plot)

grid.text("Venn Diagram of DNA and Protein Genomic Unique Regions", 
          x = 0.5,            # Horizontal position (0 = left, 1 = right)
          y = 0.98,           # Vertical position (0 = bottom, 1 = top)
          gp = gpar(fontsize = 12, fontface = "bold"))  # Font settings
```
```{r}

overlap_transcriptomic <- readr::read_tsv(file10, col_names = FALSE)
overlap_transcriptomic$ur_length <- overlap_transcriptomic$X3 - overlap_transcriptomic$X2

ggplot(overlap_transcriptomic, aes(x=ur_length)) + geom_histogram(bins = 100) +
  labs(title="Distribution of overlapping transcriptomic unique regions DNA and Protein", x="Length of overlap of unique regions", y="Frequency")+
  theme_minimal()+xlim(0,500)

```
```{r}

overlap_genomic <- readr::read_tsv(file11, col_names = FALSE)
overlap_genomic$ur_length <- overlap_genomic$X3 - overlap_genomic$X2
overlap_genomic <- overlap_genomic %>% group_by(X4) %>%
  summarize(ur_length = sum(ur_length))

ggplot(overlap_genomic, aes(x=ur_length)) + geom_histogram(bins = 100) +
  labs(title="Distribution of overlapping genomic unique regions DNA and Protein", x="Length of overlap of unique regions", y="Frequency")+
  theme_minimal()+xlim(0,500)

```